### ISSUE TITLE
Component model support

### ISSUE THREAD

**Issue by steren** [bytecodealliance/wasm-micro-runtime#2126]
WASI has moved to preview 2, which is based on the [Component Model](https://github.com/WebAssembly/component-model).

In a [recent community presentation](https://www.youtube.com/watch?v=9pLa7PUhPYA), it was briefly mentioned that the component model would be in WAMR's roadmap, but I wasn't able to find an issue or more written information about this topic. I suggest using this issue as a feature request and as a way to attach any existing links to plans and timelines for this topic.

**Comment by yamt** [comment]
technically, preview2 does not require component model.
i agree it's nice to have component model though.

btw, to make component model useful, i suspect we need to align our import/export model to the standard semantics (cf. https://github.com/bytecodealliance/wasm-micro-runtime/issues/1353)
because many of the current use cases of component model involves import/export heavily.

**Comment by lum1n0us** [comment]
> btw, to make component model useful, i suspect we need to align our import/export model to the standard semantics 

Yes, need fully alignment firstly.

**Comment by squillace** [comment]
I'm extremely interested in preview 2 and component model support in wamr for our usage. Love to hear thinking on how to help this out with wamr.

**Comment by lum1n0us** [comment]
@squillace  I am not sure about that. But would you mind sharing your usage about *preview2* and *component model*. Since all I've heard about it are kinds of concern, and you are the first one to stand for it.

**Comment by nilslice** [comment]
+1 on interest in @squillace use case! happy to try to find ways to support via contribution here.

**Comment by trcrsired** [comment]
how does preview2 work without component model? How does it know whether it is 32bit wasm or 64bit wasm?

**Comment by weiyuemin** [comment]
Any update on this? It sounds great to have component model in wamr

**Comment by lum1n0us** [comment]
There isn't a set timetable for the component model at this time. However, we are getting ready for it.  Like by restructuring the code base through the implementation of instantiation linking. We are open to any advanced plans on how to effectively support the component model and how to correctly integrate the component model into the embedded world.

**Comment by GordonSmith** [comment]
Not sure if it is of interest, but I started work on a C++ ABI implementation of the component model here:  https://github.com/GordonSmith/component-model-cpp 
I am trying to keep it runtime agnostic and am using it as an excuse to get used to some of the newer C++language features...

**Comment by lum1n0us** [comment]
Excellent. I'm looking forward to seeing an example.

**Comment by GordonSmith** [comment]
Just working through some initial samples - plenty of work todo, but hopefully you can see the use pattern take shape:
* wit:  https://github.com/GordonSmith/component-model-cpp/blob/trunk/samples/wasm/sample.wit
* wamr:  https://github.com/GordonSmith/component-model-cpp/blob/trunk/samples/wamr/main.cpp#L171-L265

**Comment by rinthnoid** [comment]
Is there any progress towards WAMR wasip2/component model since it is interesting feature yet to come in other runtimes except only wasmtime. And also the wasmtime is very hard to compile for android operating system.
It would be far more than nice if WAMR able support the component model wasm modules.

**Comment by srberard** [comment]
One thing discussed in the e-SIG was to explore using a similar approach as with JCO.  That is, preprocess the Wasm Component file into a Core Wasm module which is runnable on any Wasm runtime including WAMR (If I understand correctly, there is support changes required to the runtime as well).   Gravity did something similar with the Wazero runtime.  If there is interest in this approach, I'd be happy to set up a discussion to see what this might involve.

**Comment by rinthnoid** [comment]
Thanks @srberard for sharing this idea. As someone relatively new here, I’d definitely be interested in exploring the preprocessing approach you mentioned.

One thing I’d like to understand better: if a Wasm component is preprocessed into a core Wasm module that runs on WAMR, would the higher-level component model features (like WIT-defined functions, type safety, lists/strings handling, etc.) still be accessible to us at runtime? Or would we need to drop back to writing manual glue code in C for imports/exports and memory management?

I’m especially interested in being able to use WIT functions in a runtime other than Wasmtime, since currently Wasmtime is the only option.

I’d love to join a discussion if one is set up to go deeper into this.

**Comment by srberard** [comment]
Hi @rinthnoid, yes the full functionality of the component is available to you.  In short, the pre-processing step "unwraps" the component and does the necessary steps to run this on a Wasm Core runtime.  There are a few things that the runtime needs to export for this to work.  Of course, this is greatly simplifying the whole process.

**Comment by rinthnoid** [comment]
Thanks @srberard It would be a very nice idea to have this method, looking at the current situation of wasip2 world. So I am ready and we can setup discussion so that you we can discuss it thoroughly and it will also help users in future to use the component model with this method using other webasm runtimes.

**Comment by GordonSmith** [comment]
Getting close to a complete implementation! See https://github.com/GordonSmith/component-model-cpp/tree/WIT-CODEGEN/samples for details, now includes generated stubs from wit definitions...

**Comment by stco195** [comment]
We (at Airbus Defence and Space) are extending WAMR to support the Component Model specifications.
Currently, our focus is the Linux platform with WASI Preview 2 support, though this does not yet include the [Gated Features](https://github.com/WebAssembly/component-model/blob/main/design/mvp/Explainer.md#gated-features), like async support. We have already made some progress: parsing and instantiation are implemented, and the Canonical ABI is underway.
Our goal is to contribute this work to the upstream project via a pull request.
We are reaching out to see if anyone else in the community is working on or interested in this. We are very open to collaboration and/or exchanging ideas.

**Comment by lum1n0us** [comment]
Thanks for your excellent work. The component model is an important and significant feature. Usually, for a new feature, we merge its implementation into a development branch via some small PRs after a quick discussion about the high-level design. This approach helps reduce the impact of new features on the main branch. Once the new feature, in this case, the component model, has been fully tested and is stable, we'll merge it back into the main branch. As a starter, I'd love to create the development branch for this feature.

**Comment by lum1n0us** [comment]
[dev/cm_wasip2](https://github.com/bytecodealliance/wasm-micro-runtime/tree/dev/cm_wasip2)

**Comment by stco195** [comment]
Hi Liang,

Thank you for the support. We are happy to proceed with smaller PRs as you suggested.

Our first PR will focus on the binary parser for the component model. We have designed this to closely mirror the existing core module parser to maintain consistency with the codebase. Additionally, all new code is guarded by a -DWASM_ENABLE_COMPONENT_MODEL build flag.

Please note that we are currently completing our internal review process to release this code. It may take some time before we can push the first commit, but we are eager to get started as soon as that is cleared.

Costin

**Comment by oberstet** [comment]
rgd https://github.com/bytecodealliance/wasm-micro-runtime/tree/dev/cm_wasip2

Thanks for sharing this update! — this is exciting work and exactly what the embedded Wasm community needs.

I'm very interested in this from an MCU perspective. My target environment is *Zephyr on devices with ~512KB RAM*, where WAMR is currently the only viable runtime option. The lack of Component Model / WASIp2 support has been a significant gap, as it forces us to stay on Preview 1 while the broader ecosystem moves forward.

Beyond the technical benefits of the Component Model (explicit interfaces, composability, capability-based security), I see strong alignment with **SLSA Level 4** supply chain security requirements:

- **Hermetic builds**: Components have fully declared dependencies via WIT — no ambient authority or hidden imports
- **Provenance tracking**: Components are self-describing artifacts that can be signed, verified, and attested
- **Isolation guarantees**: The sandboxing model limits blast radius even if a component is compromised

For regulated industries (aerospace, automotive, medical, industrial) running on constrained hardware, this combination of Component Model + MCU-capable runtime could be a game-changer! I guess this won't be anything new for you, just my 2cts and positive feedback;)

**Comment by stco195** [comment]
Thank you for your patience! We have finalized our internal review process and just opened our first [PR](https://github.com/bytecodealliance/wasm-micro-runtime/pull/4872) (merge with main branch).

The binary parser will follow shortly.

**Comment by stco195** [comment]
We have open the [PR ](https://github.com/bytecodealliance/wasm-micro-runtime/pull/4889)with the CM binary parser support.

Unfortunately some of the builds in the CI/CD have failed. We are completely new to this setup and  we'd appreciate if someone could guide us what to do in this case.

**Comment by santon1972** [comment]
Hi, we have pushed all the WASM component model and WASI P2 implementation in this public Airbus forks repository branch: https://github.com/airbus-forks/wasm-micro-runtime/tree/dev/cm_wasip2_complete.
The implementation includes only the stable specifications as they were in August 2025.
The async support and wasi:http are not covered, we only prepared the implementation for future async support by introducing the Task/Subtask concepts.
The code has been tested with unit tests and WASM binaries built using both the RUST compiler and the wasi-sdk, ensuring a pretty solid code coverage.
From our perspective this project has been completed. The WAMR core team is still debating on how this feature should be integrated into the main project, so our further involvement remains to be determined.

**Comment by oberstet** [comment]
WASI: this is awesome!

> The async support and wasi:http are not covered, we only prepared the implementation for future async support by introducing the Task/Subtask concepts.

fwiw/feedback: I'm eagerly awaiting this! in my eyes, it'll be the final piece completing the picture, bring WAMR on-par with modern run-times. thanks for all the work and efforts!

**Comment by woodsmc** [comment]
My understanding is that this is one of the single largest ever code contributions to the WAMR project. My quick `cloc` analysis shows that this contribution represents 25% jump in the lines of code in the project.

Maintaining such a large code contribution for any project would be a challenge. I was wondering

1. Is there an offer from Airbus or others to help provide support for the new code contributions?
2. Genuine and related question - what's the user demand for the component model with WAMR?
    1. (and) would this translate into maintenance support?

**Background**
Looking at WAMR's recent commit history, it looks like there are maybe 2 - 3 individuals providing the bulk of support (bug & security tickets, etc.) at the moment - hence the questions about support, and the jump in project size. As with all things open source, once the new code is widely distributed and folks start testing it out, you could imagine there would be a wave of security issues / bugs / edge cases found. It would be important that the project would have the expertise of folks familiar with this addition to be able to provide the necessary support. This would mean responding to support tickets within a given number of hours / days, etc.

Seeing @oberstet 's excitement around the contribution - which is awesome. Is there an actual deployment in mind using the WAMR + Component Model, where, say, WasmTime wouldn't currently work? 

It is interesting to observe that this is the first attempt to add component model support to WAMR. The only other full runtime which has support for the component model is WasmTime. My understanding is that there hasn't been a significant heavy user of WAMR who has expressed a need for the Component Model (yet). If there was a huge demand that could be translated into maintenance support, that would be awesome.

**Comment by oberstet** [comment]
> Seeing @oberstet 's excitement around the contribution - which is awesome. Is there an actual deployment in mind using the WAMR + Component Model, where, say, WasmTime wouldn't currently work?

well, yes, that would be: porting [Crossbar.io](https://github.com/crossbario/crossbar) and [Autobahn](https://github.com/crossbario/autobahn-python), that is a [WAMP](https://wamp-proto.org/) router and client library to WAMR, including static memory allocation, to ultimately run on MCU class devices, e.g. the NXP MCX N94 and <1MB RAM. no Linux, no malloc, Zephyr and WAMR.
